Move ResourceGroup relationship into VmOrTemplate model

djberg96 opened the pull request:
This is a followup to https://github.com/ManageIQ/manageiq/pull/14000.

Originally I put the resource group relationship in the cloud_manager/vm.rb. Unfortunately, this left templates out. So, I've set the relationship in the VmOrTemplate model instead.

For backwards compatibility (and general handiness) I've created a `vms` method, as well as a `templates` method (and an :images alias). I've also updated the specs.

bdunne commented on spec/models/resource_group_spec.rb:
```diff
@@ -23,8 +23,39 @@
+    end
+
+    it "has many vms_or_templates" do
+      expect(resource_group).to respond_to(:vm_or_templates)
```
I feel like the expectation on `respond_to` is unnecessary if the method is called on the next line.

djberg96 replied:
I know people disagree with me on tests like this, but I like them because they're good for sanity checking. At worst, it's harmless.

bdunne replied:
Again, these tests appear to testing rails itself.

bdunne commented on spec/models/vm_or_template_spec.rb:
```diff
@@ -335,6 +335,19 @@
   end
 
+  context "#resource_group" do
+    before do
```
Why the `context`, `@`variables and a `before` block if there is only one test?

bdunne replied:
On second thought... Do we even need this test?  It feels like we're testing rails here.

djberg96 replied:
In case I add more tests later.

djberg96 replied:
Again, sanity testing in case someone adds scopes/filters/whatever that conflict with what I'm expecting.

jrafanie replied:
If this is a regular activerecord has_one, I don't understand why we'd need to test that.  They're usually helpful when you're developing and want to make sure you did both sides of the association correctly but aren't really needed after that.

kbrock commented on app/models/vm_or_template.rb:
```diff
@@ -79,6 +79,8 @@ class VmOrTemplate < ApplicationRecord
   has_many                  :guest_applications, :dependent => :destroy
   has_many                  :patches, :dependent => :destroy
 
+  belongs_to                :resource_group
```
:+1:

kbrock commented on app/models/resource_group.rb:
```diff
@@ -1,3 +1,8 @@
+
+  has_many :vm_or_templates
+
+  has_many :vms,       -> { where(:template => false) }, :class_name => 'VmOrTemplate'
```
Do we want to use `:class_name => 'Vm'` and `:class_name => 'Template'`?

djberg96 replied:
Without changing the tests, if I do that I get: `uninitialized constant ResourceGroup::Template`

kbrock commented:
Also, due to architectural limitations, moving up to a parent makes life simpler

kbrock commented on app/models/resource_group.rb:
```diff
@@ -1,3 +1,8 @@
 class ResourceGroup < ApplicationRecord
-  has_many :vms
+  alias_attribute :images, :templates
```
Do we need `images`?

djberg96 replied:
I just thought it would be a handy alias. :)

djberg96 commented:
@kbrock @bdunne I've removed the tests that you felt were redundant. I've also modified the `vms` method to rely on the default scope. I could not do that for the `templates` method however since there is no Template class.

bdunne commented:
@djberg96 The class is `MiqTemplate`

djberg96 commented:
@bdunne Thanks, updated.

chessbyte commented:
@bdunne @blomquisg bump
